plugin/cshared.go

```diff
@@ -0,0 +1,284 @@
+		if err != nil && !errors.Is(err, context.Canceled) {
+			fmt.Fprintf(os.Stderr, "run: %s\n", err)
+			return input.FLB_ERROR
+		}
+	}
+
+	return input.FLB_OK
+}
+
+//export FLBPluginFlush
+func FLBPluginFlush(data unsafe.Pointer, clength C.int, ctag *C.char) int {
+	initWG.Wait()
+
+	if theOutput == nil {
+		fmt.Fprintf(os.Stderr, "no output registered\n")
+		return output.FLB_RETRY
+	}
+
+	var err error
+	once.Do(func() {
+		runCtx, runCancel = context.WithCancel(context.Background())
+		theChannel = make(chan Message, 1)
+		go func() {
+			err = theOutput.Flush(runCtx, theChannel)
+		}()
+	})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "run: %s\n", err)
+		return output.FLB_ERROR
+	}
+
+	select {
+	case <-runCtx.Done():
+		err := runCtx.Err()
+		if err != nil && !errors.Is(err, context.Canceled) {
+			fmt.Fprintf(os.Stderr, "run: %s\n", err)
+			return output.FLB_ERROR
+		}
+
+		return output.FLB_OK
+	default:
+	}
+
+	in := C.GoBytes(data, C.int(clength))
+	h := &codec.MsgpackHandle{}
+	err = h.SetBytesExt(reflect.TypeOf(bigEndianTime{}), 0, &bigEndianTime{})
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "big endian time bytes ext: %v\n", err)
+		return output.FLB_ERROR
+	}
+
+	dec := codec.NewDecoderBytes(in, h)
+
+	for {
+		select {
+		case <-runCtx.Done():
+			err := runCtx.Err()
+			if err != nil && !errors.Is(err, context.Canceled) {
+				fmt.Fprintf(os.Stderr, "run: %s\n", err)
+				return output.FLB_ERROR
+			}
+
+			return output.FLB_OK
+		default:
+		}
+
+		var entry []any
+		err := dec.Decode(&entry)
+		if errors.Is(err, io.EOF) {
+			break
+		}
+
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "decode: %s\n", err)
+			return output.FLB_ERROR
+		}
+
+		if d := len(entry); d != 2 {
+			fmt.Fprintf(os.Stderr, "unexpected entry length: %d\n", d)
+			return output.FLB_ERROR
+		}
+
+		ft, ok := entry[0].(bigEndianTime)
+		if !ok {
+			fmt.Fprintf(os.Stderr, "unexpected entry time type: %T\n", entry[0])
+			return output.FLB_ERROR
+		}
+
+		t := time.Time(ft)
+
+		recVal, ok := entry[1].(map[any]any)
+		if !ok {
+			fmt.Fprintf(os.Stderr, "unexpected entry record type: %T\n", entry[1])
+			return output.FLB_ERROR
+		}
+
+		var rec map[string]string
+		if d := len(recVal); d != 0 {
+			rec = make(map[string]string, d)
+			for k, v := range recVal {
+				key, ok := k.(string)
+				if !ok {
+					fmt.Fprintf(os.Stderr, "unexpected record key type: %T\n", k)
+					return output.FLB_ERROR
+				}
+
+				val, ok := v.([]uint8)
+				if !ok {
+					fmt.Fprintf(os.Stderr, "unexpected record value type: %T\n", v)
+					return output.FLB_ERROR
+				}
+
+				rec[string(key)] = string(val)
+			}
+		}
+
+		tag := C.GoString(ctag)
+		// C.free(unsafe.Pointer(ctag))
+
+		theChannel <- Message{Time: t, Record: rec, tag: &tag}
+
+		// C.free(data)
+		// C.free(unsafe.Pointer(&clength))
+	}
+
+	return output.FLB_OK
+}
+
+//export FLBPluginExit
+func FLBPluginExit() int {
+	log.Printf("calling FLBPluginExit(): name=%q\n", theName)
+
+	if runCancel != nil {
```

Comment:
will this ever be nil?

Reply:
I don't know how the whole plugin lifecycle works from C side. I thought the plugin could be registered and exit for some reason right after. Or even initialized without any collection/flush and exited right after.
That's why I put it here to be safe in the case if was not initialized.